Apps/FlightSoftware/Watchdog2/src/i2c.c

```diff
@@ -0,0 +1,318 @@
+#include "include/i2c.h"
+
+// TODO: integrate this code with the rest of the system
+
+int8_t raw_battery_charge[2];
+int8_t raw_battery_voltage[2];
+int8_t raw_battery_current[2];
+int8_t raw_fuel_gauge_temp[2];
+
+uint8_t batt_charge_telem;
+uint8_t batt_curr_telem;
+
+uint8_t fuel_gauge_write_control_reg;
+uint8_t fuel_gauge_read_control_reg;
+
+/*
+ * File for interfacing with I2C protocol hardware module
+ *
+ * I2C communication uses eUSCI_B0 (power circuitry)
+ */
+
+/* Used to track the state of the software state machine*/
+I2C_Mode MasterMode = IDLE_MODE;
+
+/* The Register Address/Command to use*/
+uint8_t TransmitRegAddr = 0;
+
+// initialize buffers and their corresponding byte counters
+uint8_t ReceiveBuffer[I2C_RX_BUFFER_MAX_SIZE] = {0};
+uint8_t RXByteCtr = 0;
+uint8_t ReceiveIndex = 0;
+uint8_t TransmitBuffer[I2C_TX_BUFFER_MAX_SIZE] = {0};
+uint8_t TXByteCtr = 0;
+uint8_t TransmitIndex = 0;
+
+/* uart0 rx handler (will only be called after SLIP packet fully received) */
+void (*i2c_rx_handler)(uint16_t len, struct buffer *buf) = (void *)(0);
+
+
+/* init function */
+void i2c_init() {
+    // Configure i2c interface
+    P1SEL1 |= BIT6; // P1.6 SDA
+    P1SEL1 |= BIT7; // P1.7 SCL
+
+    // pins for using eusci_b2 (only ones available on launchpad)
+//    P7SEL0 |= BIT0 | BIT1;
+//    P7SEL1 &= ~(BIT0 | BIT1);
+
+    UCB0CTLW0 = UCSWRST;                      // Enable SW reset
+    UCB0CTLW0 |= UCMODE_3 | UCMST | UCSSEL__SMCLK | UCSYNC; // I2C master mode, SMCLK
+    UCB0BRW = 160;                            // fSCL = SMCLK/160 = ~100kHz
+    UCB0I2CSA = I2C_SLAVE_ADDR;                   // Slave Address
+    UCB0CTLW0 &= ~UCSWRST;                    // Clear SW reset, resume operation
+    UCB0IE |= UCNACKIE;
+}
+
+
+
+/* For slave device with dev_addr, read the data specified in slaves reg_addr.
+ * The received data is available in ReceiveBuffer
+ *
+ * dev_addr: The slave device address.
+ *           Example: I2C_SLAVE_ADDR
+ * reg_addr: The register or command to send to the slave.
+ *           Example: CMD_TYPE_0_SLAVE
+ * count: The length of data to read
+ *           Example: TYPE_0_LENGTH
+ *  */
+I2C_Mode I2C_Master_ReadReg(uint8_t dev_addr, uint8_t reg_addr, uint8_t count)
+{
+    /* Initialize state machine */
+    MasterMode = TX_REG_ADDRESS_MODE;
+    TransmitRegAddr = reg_addr;
+    RXByteCtr = count;
+    TXByteCtr = 0;
+    ReceiveIndex = 0;
+    TransmitIndex = 0;
+
+    /* Initialize slave address and interrupts */
+    UCB0I2CSA = dev_addr;
+    UCB0IFG &= ~(UCTXIFG + UCRXIFG);       // Clear any pending interrupts
+    UCB0IE &= ~UCRXIE;                       // Disable RX interrupt
+    UCB0IE |= UCTXIE;                        // Enable TX interrupt
+
+    UCB0CTLW0 |= UCTR + UCTXSTT;             // I2C TX, start condition
+
+    __bis_SR_register(GIE);              // Enable interrupts
+    __delay_cycles(10000);               // give fuel gauge time to respond
```

Comment:
It's not clear to me what this delay actually achieves, since all we do after it is return a fixed value. Can you explain why this is needed?

Reply:
@SchwartzCode

Reply:
Ok, after some digging I think I see where a lot of the issues in this file are coming from. This file is clearly "based" on [this code](https://canvas.northwestern.edu/courses/92393/files/6968257/download?download_frd=1) (which may be fine, I don't see any copyright claim in that file's header). However, there are a few major differences that make some of the code in that file potentially unsafe. First of all, in that file they go into LPM at the end of `I2C_Master_ReadReg()` and `I2C_Master_WriteReg()`, just before the return statement. LPM isn't exited until the ISR exits following the completion of a read or write, so there is no contention for any of the variables that are used in both the ISR and non-ISR functions. This is why that file got away without declaring those shared variables as volatile. This file kind of emulates entering LPM with these big sleeps, as they also make it so there is no contention between the use of the shared variables in the ISR and out of it. However, if you're just gonna sleep() here you may as well enter LPM, because you're not doing anything in the sleep anyway. Entering LPM has the added benefits of being low power (obviously), but also means that completing the read or write is signaled by the interrupt that completes the action (rather than a presumption that the action will be complete by the time this sleep completes). However, I imagine an issue that was run into was that ISRs in other watchdog code caused an exit from LPM, and that meant that the original code in the linked file didn't work when slapped into this use. This means that there's no good way to make the original approach of the code in the linked file work without a big sleep like this.

I'm working on writing up some code for an entirely different approach that will hopefully work without needing to do these long sleeps. However, we can't be sure my code will work, it'll take me a little bit to write it, and we have very little time to vet it. We'll just have to play it by ear and see if it works out.

Reply:
Hi Matt, I got the code from [TI's site](https://dev.ti.com/tirex/explore/node?node=AKcP2HOBL6serErY28eqNw__IOGqZri__LATEST), which I'd wager the file you linked is also based on (or vice-versa). 

You are spot on with the LPM/other ISRs issue (I think), we had issues with ISR traps when we used the LPM so that's why we switched to the delays. Thank you for trying to write a different version! I was at a loss trying to think of a new way to do this

Reply:
For anyone that doesn't already know, my new code is in a separate PR here: https://github.com/PlanetaryRobotics/CubeRoverPackage/pull/59